src/main/java/com/google/sps/servlets/SearchMealServlet.java

```diff
@@ -0,0 +1,77 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.gson.Gson;
+
+@WebServlet("/search-meal")
+public class SearchMealServlet extends HttpServlet {
+
+    private ArrayList<String> dishes;
+    private ArrayList<String> result;
+
+    @Override
+    public void init() {
+        dishes = new ArrayList<>();
+        dishes.add("fried potato with mushrooms and onion");
+        dishes.add("pizza with pineaple, sausage and tomato");
+        dishes.add("vegetable soup with onion");
+        dishes.add("chocolate cake with butter cream and strawberry");
+        dishes.add("hot chocolate with sugar and caramel");
+    }
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String gson = new Gson().toJson(result);
```

Comment:
Please make `result` a local variable in the function(s) you use.

Here you're getting the `result` variable, which might be empty, or might be different depending on what kind of POST request was called beforehand. You have no guarantee which of the two will happen, because the servlet may be restarted in the meantime - especially on AppEngine, where under heavy load multiple instances of this servlet will be started, each running independently, and the user's request may hit a different instance every time.

Reply:
Well, I dont exactly understand how I have to keep data about result of search.

Reply:
The suggestion here follows what @bprusak suggested above in the comment regarding RESTful services.

Long story short, if I understand correctly, the suggestion is to use only `doGet` to perform the search. It would require removing the current `doPost` method completely. This way you wouldn't need to store `result` in a private variable, because it would be both created and used in the same method (here, `doGet`).

In general, the less private properties representing some kinda state a class has, the less likely it is to have a bug. Managing state is tricky, and the easiest way of solving the problem is by avoiding it.

Let's take a look at the following example. From what I understand, there's an assumption now that a user performs a `POST` request which populates the `result` variable, and soon after the same user performs a `GET` request, which fetches the stored data.
What would happen if we have 2 users using the app at the same time?
1. User A sends a `POST`, searching for "chocolate". We set `result= ["hot chocolate with sugar and caramel"]` and return a OK response.
2. Soon after that, User B sends a `POST` searching for "potato". We set `result=["fried potato with mushrooms and onion"]`.
3. User A wants to retrieve his search results, expecting some chocolate-related suggestions. We encode `result` and return it to the user. However, in step 2. we set `result` to a new value: french fries, and that's the result User A receives.

If we didn't store the temporary state in a private variable but kept it local (by doing both fetching and returning the result in the same request), no such problems would appear.

Reply:
I will resolve it in the next steps when I will create REST API